File: action_controller/src/action_controller/actioncontroller.py
#! /usr/bin/env python

## @package action_controller
# Listens to trajectory and sends it spiri controller
# @author Rohan Bhargava
# @version 1.1.1
import roslib

#roslib.load_manifest('actionlib_tutorials')
roslib.load_manifest('action_controller')

import rospy
import actionlib
import actionlib_tutorials.msg
import moveit_msgs.msg
from actionlib.exceptions import *

import action_controller.msg
import trajectory_msgs.msg
from geometry_msgs.msg import Transform
from geometry_msgs.msg import Twist, PoseStamped,Pose
from nav_msgs.msg import Odometry
import time


from spiri_api import pid
from collections import deque

## Class to listen for trajectories from moveit_simple_controller manager.
class actioncontroller(object):
  ## Constructor
	def __init__(self,name):

		self.state=Pose()
		self._action_name=name
		self.traj=trajectory_msgs.msg.MultiDOFJointTrajectory()
		self._as=actionlib.ActionServer(self._action_name,ActionSpec=action_controller.msg.MultiDofFollowJointTrajectoryAction,goal_cb=self.goalcb,cancel_cb=self.cancelcb,auto_start=False)
		self.pub=rospy.Publisher('command',trajectory_msgs.msg.MultiDOFJointTrajectory,queue_size=1)
		rospy.Subscriber('/ground_truth/state',Odometry,self.callback)
		self.lastPosition=Transform()
		self.vel=Twist()
		self.flag=True
		self.has_active_goal=False
		self.active_goal=actionlib.ServerGoalHandle()
		self._as.start()
		
		
	## Callback function for the state topic. This function is just used so that we can inform move it that execution is completed. Just a hack
	# @param self Object pointer
	# @param data Contains data published on the topic
	def callback(self,data):

		
		
		if rospy.get_param('/has_active_goal',True)==True:
		  return
		if rospy.get_param('/execution_completed',False)==False:
		  return
		  
		self.active_goal.set_succeeded()
		rospy.set_param('/execution_completed',False)
		
	## Function is called whenever a new goal has been recieved. 
	# @param gh Goal handle
	def goalcb(self,gh):
		

		if(gh.get_goal().trajectory.joint_names[0]!='virtual_join'):
		    rospy.logerr('Joints on incoming goal dont match our joints')
		    gh.set_rejected()
		    return
		
		gh.set_accepted('goal accepted')
	        self.traj=gh.get_goal().trajectory
		self.active_goal=gh
		rospy.set_param('/has_active_goal',True)
		self.pub.publish(self.traj)
	## Function is called whenever a goal has been cancelled. 
	# @param gh Goal handle	
	def cancelcb(self,gh):
		
		if(self.active_goal==gh):
			
			rospy.set_param('/has_active_goal',False)
			


	
if __name__ == '__main__':

	rospy.init_node('multi_dof_joint_trajectory_action')
	#print rospy.get_name()

	actioncontroller(rospy.get_name())
	rospy.spin()

File: action_controller/src/actioncontroller.cpp
/*!
  \package action_controller
  \author Rohan Bhargava
  \version 1.1.2
  */

#include <ros/ros.h>
#include <actionlib/server/action_server.h>
#include <pthread.h>
#include <action_controller/MultiDofFollowJointTrajectoryAction.h>
#include <geometry_msgs/Twist.h>
#include <trajectory_msgs/MultiDOFJointTrajectory.h>
#include <nav_msgs/Odometry.h>
/*!
 * \brief This class listens for trajectories from moveit_simple_controller manager
 */
class Controller{
private:
    typedef actionlib::ActionServer<action_controller::MultiDofFollowJointTrajectoryAction> ActionServer;
    typedef ActionServer::GoalHandle GoalHandle;

public:
    /*!
     * \brief Constructor
     * \param n Handle for the node
     */
    Controller(ros::NodeHandle &n):
        node_(n),action_server_(node_,"multi_dof_joint_trajectory_action",boost::bind(&Controller::goalCB,this,_1),boost::bind(&Controller::cancelCB,this,_1),false),has_active_goal(false)
    {
      action_server_.start();
      ROS_INFO("Node ready");

    pub=node_.advertise<trajectory_msgs::MultiDOFJointTrajectory>("/command",1);
    sub=node_.subscribe<nav_msgs::Odometry>("/ground_truth/state",1000,boost::bind(&Controller::callback,this,_1));
    }
private:
    trajectory_msgs::MultiDOFJointTrajectory_<std::allocator<void> > traj;
    GoalHandle active_goal;
    ros::NodeHandle node_;
    ActionServer action_server_;
    ros::Publisher pub;
    ros::Subscriber sub;
    bool has_active_goal;
    bool execution_completed;
    /*!
     * \brief Functions is called whenever a new goal has been received
     * \param gh Goal handle
     */


    void goalCB(GoalHandle gh)
    {
        gh.setAccepted();
        traj=gh.getGoal()->trajectory;
        active_goal=gh;

        ros::param::set("/has_active_goal",true);
        pub.publish(traj);


    }
    /*!
     * \brief Functions is called whenever a goal has been cancelled
     * \param gh Goal handle
     */
    void cancelCB(GoalHandle gh)
    {
        if(active_goal==gh)
        {
            ros::param::set("/has_active_goal",false);
        }
    }
    /*!
     * \brief Callback functions for the state topic.
     * \todo This function is just used so that we can inform move it that executions is completed. Is it a hack?
     * \param msg Contains data published on the topic
     */
    void callback(const nav_msgs::OdometryConstPtr& msg)
    {
        ros::param::param<bool>("/has_active_goal",has_active_goal,1);
        if(has_active_goal==1)
        {
        return;
        }
        ros::param::param<bool>("/execution_completed",execution_completed,0);
        if( execution_completed==0)
        {
            return;
        }
        active_goal.setSucceeded();
        ros::param::set("/execution_completed",false);


    }
};
int main(int argc, char** argv)
{
    ros::init(argc, argv, "spiri_controller");
    ros::NodeHandle node;//("~");
    Controller control(node);

    ros::spin();

    return 0;
}


